### ISSUE TITLE
Getting the Texture ID from a vispy.gloo.Texture object

### ISSUE THREAD

**Issue by Patrick-Dietrich** [vispy/vispy#1027]
For the interoperation of OpenGL and OpenCL it is necessary to get the texture ID from a vispy.gloo.Texture object. Getting the IDs of other OpenGL objetcs like VBOs would be equally usefull.

**Comment by Eric89GXL** [comment]
Agreed. @almarklein @rossant any idea how this should work with GLIR?

**Comment by rossant** [comment]
I don't see how it could be possible but @almarklein might have better ideas

**Comment by campagnola** [comment]
We've discussed this a few times in various places. Right now gloo/glir implement a one-way communication to the opengl context, making it impossible (or at least very inconvenient) to access any return values. What we need is to make gloo into a proper asynchronous communication API--all function calls should return a Future, and the user can optionally block until the return value becomes available.

**Comment by bollu** [comment]
@campagnola I'm planning on re-implementing GLIR in C++ for [a game I'm writing for funsies](https://github.com/bollu/lovecraft). Could you detail the design you have in your head a little bit? It would be an interesting exercise.

**Comment by campagnola** [comment]
@bollu, we've already been discussing the eventual need to reimplement some of vispy's core pieces in a faster language (following @rossant's vision here: http://cyrille.rossant.net/compiler-data-visualization/). If you're up for this, there's a good chance it will be useful to us in the future.

Basically, there is no way for us to access return values from GL calls. Initially, this was deliberate because when a gloo function is called, there may not yet be any GL context available, and even if there is a context, we don't want to send every return value back to the caller in the case of a webgl backend. In some cases, however, we really do need these return values, and we are happy to wait until a context is available to get them. What we'd like, then, is an API like https://docs.python.org/3/library/concurrent.futures.html#future-objects:

```
future = gloo.some_function()
...
return_val = future.result()  # block until return value is available.
```

Calling `future.result()` would immediately flush the GLIR cache and grab the return value (possibly waiting if the context is remote). If no context is available at all, then calling `result()` would just raise an exception.

Usually we are not interested in return values, so to avoid the overhead of dealing with unneeded returns, most functions will have returns disabled by default (and will not create a Future). Functions that create new objects or return useful values (all of the `getXXX` functions) will generate a Future by default. Probably we can override this behavior using a keyword arg to the function.

**Comment by bollu** [comment]
I read @rossant's blogpost, as well as your response. These are my thoughts, based on what little I know about graphics, compilers and LLVM.

>What would it take to make this use-case a reality?

>There are several components:

>1. A compiler API to specify interactive visualizations programmatically and compile them to platform-independent files

>2. A Vulkan runtime that executes visualizations on the GPU

>3. Optionally, a CPU runtime that executes visualizations on the CPU, as a fallback if a GPU is not available

1. LLVM -> SPIR-V is pretty damn awesome. I wasn't aware of this. However, I'm quite hesitant about the performance implications of this. LLVM as an IR isn't particularly "suited" (AFAIK) to model the way the GPU works, since its primitives are more programming language oriented. [Here's an argument that LLVM is a poor choice of format to "transmit" program information](http://lists.cs.uiuc.edu/pipermail/llvmdev/2011-October/043719.html)

2. what does "compiler API" mean? AFAIK, we'd have a hypothetical language that compiler down to SPIR-V. So what we need is a compiler. Where did the API bit come into play?

3. What does a "Vulkan Runtime" mean? Isn't Vulkan just a lower-level replacement to OpenGL? There is no "runtime" involved, (unless you consider the drivers to be the runtime..?) you just create command buffers and send data and code to the GPU.

4. A CPU runtime would be very expensive and would have terrible performance IMO (like the way the old Intel chipsets had software shaders).

5. Another opinion of mine is that C++11 is nice, but it is __huge__. The addition of move semantics plus a bunch of other features into the language just makes it that much larger. I'm all for C++, but, what are your thoughts on, say, Rust?

6. Interesting that Python promises sort of prioritize synchronization, while [JavaScript promises](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise) are always async. 

I would personally like to have an API where you could `peek` a `Promise` to see if it is complete or not (so it could be used in event-loop like situations). You could use `then` to chain async function calls, and a `force` to force a synchronization. 

As for the getters, Haskell has a nice convention where `<func_name>_` is the one with no return values, while `<func_name>` would have a useful return. Examples: `replicateM_`, `sequence_`, ...

Anyway, I'm probably getting way too specific when it comes to the API

7. how does picking work right now in VisPy if you don't have return values in `gloo`?

**Comment by campagnola** [comment]
> how does picking work right now in VisPy if you don't have return values in gloo?

It doesn't :)
On desktop we bypass gloo, and on webgl we have a very limited workaround that does 2D bounding-box intersection to allow viewboxes to interact.

**Comment by bollu** [comment]
Ah, I didn't know that about gloo :)

Also, another thing about the Implementation - IMO, we should have a pure C
core, which is wrapped by C++11 niceties.

Binding C++ to any other language is waaay harder than binding C. C FFI for
most languages is considerably easy. C++ is not at all so.

**Comment by rossant** [comment]
> LLVM -> SPIR-V is pretty damn awesome. I wasn't aware of this. However, I'm quite hesitant about the performance implications of this. LLVM as an IR isn't particularly "suited" (AFAIK) to model the way the GPU works, since its primitives are more programming language oriented. Here's an argument that LLVM is a poor choice of format to "transmit" program information

regarding LLVM/SPIR-V, you'll find interesting dicussions on the LLVM mailing list

about whether LLVM is adapted or not: the only way to know is to try ;) that's why it's such an experimental project: there are many ways it can go wrong. But I definitely thing it's worth trying.

> what does "compiler API" mean? AFAIK, we'd have a hypothetical language that compiler down to SPIR-V. So what we need is a compiler. Where did the API bit come into play?

i meant a backend compiler API. there would be no hypothetical language apart from Python GLSL etc. (anything that can compile to LLVM)

> What does a "Vulkan Runtime" mean? Isn't Vulkan just a lower-level replacement to OpenGL? There is no "runtime" involved, (unless you consider the drivers to be the runtime..?) you just create command buffers and send data and code to the GPU.

I just thought about a Vulkan wrapper in C++, perhaps the term of runtime is not good but I could not find a better alternative! you would call this wrapper at runtime

> A CPU runtime would be very expensive and would have terrible performance IMO (like the way the old Intel chipsets had software shaders).

I don't think it would be that terrible with compiled code and multithreading (perhaps 10x slower or something). In any case this is not meant to provide performance but rather a fallback, testing tool, etc

> Another opinion of mine is that C++11 is nice, but it is huge. The addition of move semantics plus a bunch of other features into the language just makes it that much larger. I'm all for C++, but, what are your thoughts on, say, Rust?

I thought about C and Rust too. Rust is highly interesting but I'm afraid it's too new and not mature compared to C++ 
C is extremely low level and much less safe than C++. You need to be a very experienced programmer in C to write good code, which i am not. C++11, if you restrict yourself to the most basic constructs, the most widely used parts of stdlib etc, is easier to work with IMO. The book by scott meyers on modern c++ is excellent in this respect

**Comment by bollu** [comment]
> there would be no hypothetical language apart from Python GLSL etc. 

Ah, that clears up a lot :) So, there'll be let's say Python code, that gets compiled down to LLVM IR, which then gets compiled/transpiled to SPIR-V?

> I just thought about a Vulkan wrapper in C++, perhaps the term of runtime is not good but I could not find a better alternative! you would call this wrapper at runtime

Yep, makes sense

> I just thought about a Vulkan wrapper in C++,In any case this is not meant to provide performance but rather a fallback, testing tool, etc

Right, but, what are your thoughts on the FFI? I don't think Python supports the C++ calling convention (`_thiscall`) for the FFI. Only C-style calling convention is supported IIRC.

So, even if the library is written in C++, there would still be a need for C wrappers, right?


> C is extremely low level and much less safe than C++

Yeah, that is very true. Manual memory management makes it hard to be sure about leaks / dangling pointers.

Just to raise interesting possibilities, what are your thoughts on [Jai](https://sites.google.com/site/jailanguageprimer/)?

**Comment by the-lay** [comment]
Is there still no way to obtain texture handle from gloo texture? And if so, how should I do with Vispy, or should I use PyOpenGL for texture creation, and then somehow pass it back to gloo.Program as a uniform?

My goal is to do some operations with PyCUDA and write the result to a 3d texture that will be shown by Vispy.

Sorry for necroposting.
